### ISSUE TITLE
lateinit variable doesn't work with interceptTestCase

### ISSUE THREAD

**Issue by lenguyenthanh** [kotlintest/kotlintest#174]
I have a problem with init `lateinit var` property in `interceptTestCase` method.
Here is my example:
```
class ExampleTest : StringSpec() {
    private lateinit var someString: String

    override fun interceptTestCase(context: TestCaseContext, test: () -> Unit) {
        println("Start Testcase")
        someString = "Hello"
        test()
        println("End Testcase")
    }

    init {
        "Hello should equal to Hello" {
            someString shouldEqual "Hello"
        }
    }
}
```

And here is out put:
```
Start Testcase

kotlin.UninitializedPropertyAccessException: lateinit property someString has not been initialized

	at com.locata.myapplication.ExampleTest.access$getSomeString$p(ExampleTest.kt:7)
	at com.locata.myapplication.ExampleTest$1.invoke(ExampleTest.kt:18)
	at com.locata.myapplication.ExampleTest$1.invoke(ExampleTest.kt:7)
	at io.kotlintest.Spec$runTest$callable$1$1.invoke(Spec.kt:124)
	at io.kotlintest.Spec$runTest$callable$1$1.invoke(Spec.kt:15)
	at io.kotlintest.Spec$runTest$initialInterceptor$1$1.invoke(Spec.kt:116)
	at io.kotlintest.Spec$runTest$initialInterceptor$1$1.invoke(Spec.kt:15)
	at com.locata.myapplication.ExampleTest.interceptTestCase(ExampleTest.kt:12)
	at io.kotlintest.Spec$runTest$initialInterceptor$1.invoke(Spec.kt:116)
	at io.kotlintest.Spec$runTest$initialInterceptor$1.invoke(Spec.kt:15)
	at io.kotlintest.Spec$runTest$callable$1.call(Spec.kt:124)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)


Process finished with exit code 255
```

It's happen for both 2.0.0 and 2.0.1.

**Comment by lenguyenthanh** [comment]
Just investigated this issue. In example above, We use oneInstancePerTest so the `someString` in test is in different instance with `someString` we initialized in `interceptTestCase`

So if we use 
```
    override val oneInstancePerTest: Boolean
        get() = false```

or 
```
    override fun interceptTestCase(context: TestCaseContext, test: () -> Unit) {
        (context.spec as LateInitValuesTest).lateInitString = "Hello"
        test()
    }```

Our test will work as expected.

**Comment by helmbold** [comment]
Thanks for your investigation. I think we can close this issue than or do you have any objections?

By the way: Overriding a val can be a bit more concise:

    override val oneInstancePerTest = false

**Comment by lenguyenthanh** [comment]
I think we can point this issue in our document and then close this issue

**Comment by lenguyenthanh** [comment]
I just wonder why we set `oneInstancePerTest` to be True as default? And do we have any solution for this problem? I think it causes a lot of confuses for new comers.

**Comment by helmbold** [comment]
`oneInstancePerTest` is true by default, because it effectivly isolates tests from each other and is thus a safer default.

**Comment by lenguyenthanh** [comment]
Thanks @helmbold for explanation.

**Comment by panell** [comment]
This is really big issue. 
The proposed solution works with StringSpec
```
class MyTests : BehaviorSpec() {
    lateinit var profilePresenter: ProfilePresenter
    @Mock lateinit var mockEventBus: EventBus

    override val oneInstancePerTest = false

    override fun interceptTestCase(context: TestCaseContext, test: () -> Unit) {
        MockitoAnnotations.initMocks(this)

        profilePresenter = ProfilePresenter(mockEventBus)

        test() // don't forget to call test()!
    }

    init {
        "Should register event buss" {
            profilePresenter.takeView(mockView)
            verify(mockEventBus).register(profilePresenter)
        }
    }
}
```

but it fails to work with BehaviorSpec
```
class MyTests : BehaviorSpec() {
    lateinit var profilePresenter: ProfilePresenter
    @Mock lateinit var mockEventBus: EventBus

    override val oneInstancePerTest = false

    override fun interceptTestCase(context: TestCaseContext, test: () -> Unit) {
        MockitoAnnotations.initMocks(this)

        profilePresenter = ProfilePresenter(mockEventBus)

        test() // don't forget to call test()!
    }

    init {
        Given("ProfilePresenter") {

            When("take view") {
                profilePresenter.takeView(mockView)

                Then("register event bus") {
                    verify(mockEventBus).register(profilePresenter)
                }
            }
        }
    }
}
```

kotlin.UninitializedPropertyAccessException: lateinit property profilePresenter has not been initialized

**Comment by lawrenceching** [comment]
It's very unexpected. From your doc I think `oneInstancePerTest` should be false by default.

> One Instance Per Test (since 1.1.0)
>By default a single instance of the test class is created for all the test it contains. However, if you wish to have a fresh instance per test (sometimes its easier to have setup code in the init block instead of resetting after each test) then simply override the oneInstancePerTest value and set it to true

I would suggest to describe much clearly especially for the JUnit user(who always use @Before annotation).

**Comment by helmbold** [comment]
After thinking a bit about it, I'm wondering why the `interceptTestCase` method isn't executed on the same instance a the test method. Maybe we could fix that in the [runTest](https://github.com/kotlintest/kotlintest/blob/master/src/main/kotlin/io/kotlintest/Spec.kt#L106) method. Choosing a safer default would hide the problem, but it wouldn't solve it.

**Comment by michalbrz** [comment]
I looked into this issue and the thing that is important is **on which instance of Spec class** `test` and `interceptTestCase` methods are called.

If configuration is the default - `oneInstancePerTest` set to true - indeed one instance is created for every test, as you would expect. But `interceptTestCase` is called on one base instance of Spec that is creating all other instances. Thus, `test` and `interceptTestCase` are never called on the same instance so they never share fields.

In comparison, if `oneInstancePerTest` is set to false, then **always** the same instance is used, so naturally `test` and `interceptTestCase` share fields. Therefore, in this case changing field value in `oneInstancePerTest` influences field value in `test`.

It seems to me it doesn't work as intended, so I fixed it and created PR: https://github.com/kotlintest/kotlintest/pull/230

**Comment by helmbold** [comment]
Thanks for converting my vague hint in code!

**Comment by tonilopezmr** [comment]
Please solve this problem!

**Comment by antoniolg** [comment]
In case it helps, based on the second proposal here https://github.com/kotlintest/kotlintest/issues/174#issuecomment-300058419, I'm just wrapping everything in a `with` block that uses the `context.spec` casted to the right type:

```java
    override fun interceptTestCase(context: TestCaseContext, test: () -> Unit) {
        with(context.spec as NotificationsDaoTest){

            database = ...
            notificationsDao = database.notificationDao()

            test()

            database.close()
        }
    }
```

Thanks for the comments here, they saved my day. I'm finding interceptors a bit counterintuitive. Maybe it's becase I'm used to `@Before` and `@After`...

**Comment by sksamuel** [comment]
This is fixed in 3.0.0 RC1.
If you have the time, please give it a try. There are some breaking changes that should be aware of by reading the changelog here: https://github.com/kotlintest/kotlintest/blob/master/CHANGELOG.md

**Comment by tonilopezmr** [comment]
well done @sksamuel 👏
